src/registry.ts

```diff
@@ -0,0 +1,41 @@
+import { Document } from "@govtechsg/open-attestation";
+import { ethers, Signer } from "ethers";
+import { getBatchMerkleRoot } from "./util/token";
+import { abi as TokenRegistryABI } from "../build/contracts/ERC721MintableFull.json";
+
+export class TokenRegistry {
+  web3Provider: Signer;
+
+  address: string;
+
+  contractInstance: any;
+
+  /**
+   * Creates a TokenRegistry instance with the specified address and ethersjs signer
+   * @param {object} instanceParameters
+   * @param {string} instanceParameters.contractAddress - Address that the TokenRegistry is located at
+   * @param {Signer} instanceParameters.web3Provider - An ethers.js signer instance with read/write access
+   */
+  constructor({ contractAddress, web3Provider }: { contractAddress: string; web3Provider: Signer }) {
+    this.web3Provider = web3Provider;
+    this.address = contractAddress;
+    // doing JSON.stringify below because of ethers.js type issue: https://github.com/ethers-io/ethers.js/issues/602
+    this.contractInstance = new ethers.Contract(contractAddress, JSON.stringify(TokenRegistryABI), web3Provider);
+  }
+
+  async mint(document: Document, owner: string) {
+    const tokenId = getBatchMerkleRoot(document);
+    return this.contractInstance["safeMint(address,uint256)"](owner, tokenId);
+  }
+
+  async ownerOf(document: Document) {
+    const tokenId = getBatchMerkleRoot(document);
+    return this.contractInstance.ownerOf(tokenId);
+  }
+
+  async transferTo(document: Document, newOwner: string) {
+    const tokenId = getBatchMerkleRoot(document);
+    const currentOwner = await this.ownerOf(document);
+    return this.contractInstance["safeTransferFrom(address,address,uint256)"](currentOwner, newOwner, tokenId);
```

Comment:
does it look like :hankey: because there are multiple `safeTransferFrom` with different signature ?

Reply:
exactly